deepchem/data/datasets.py

```diff
@@ -644,50 +665,82 @@ def iterate(dataset):
         shard_perm = np.random.permutation(num_shards)
       else:
         shard_perm = np.arange(num_shards)
-      pool = Pool(1)
-      next_shard = pool.apply_async(dataset.get_shard, (shard_perm[0],))
+
+      n_workers = min(4, num_shards)
+
+      # (ytz): Depending on the application, thread-based pools may be faster
+      # than process based pools, since process based pools need to pickle/serialize
+      # objects as an extra overhead. Also, as hideously as un-thread safe this looks,
+      # we're actually protected by the GIL.
+      pool = ThreadPool(n_workers)
```

Comment:
@peastman I modified your multiprocessing pool to use ThreadPools instead. I'm not sure if this is ideal. In my case I was facing perf overhead associated with pickling over IPC (interprocess communication), so I changd this to a ThreadPool. I also modified the number of workers I use in parallel.

Reply:
It was already using threads, not processes.  `multiprocessing.dummy` is a complete implementation of the `multiprocessing` API, only implemented with threads instead of processes.  Unlike ThreadPool, it's an officially documented and supported API.

Also, I specified only one worker thread for a reason.  You don't want it trying to load every shard in parallel all at once!  That would defeat the whole purpose of sharding in the first place.  With my version, there's never more than two shards in memory at once: the one it's currently processing, and the next one that it's preloaded.

Reply:
This loads at max_workers number of threads. On the ThreadPool side, I had no idea that dummy was actually a wrapper around it, but with a modicum more of documentation. I personally found ThreadPool to be more obvious (that it's using threads), but I can revert it back to dummy if you'd like.

https://stackoverflow.com/questions/3033952/threading-pool-similar-to-the-multiprocessing-pool

Reply:
That's fine as long as there's a large number of small shards.  If there's a small number of large shards, this could use a lot more memory than people are expecting.  I felt a little bad about even having two shards in memory at once, but this version could have up to five in memory at once.